kafka: add production docs (#101)
kudobuilder/frameworks: repository/kafka/docs/latest/production.md

![kudo-kafka](./resources/images/kudo-kafka.png)

# Running KUDO Kafka in production

## Checklist

- Verify the storage class features
- Broker configuration
- Run the KUDO Kafka with tuned parameters

> This document gives preference to the **data durability over the performance**.

Reviewer: ```suggestion
This document optimizes for **data durability over performance**. 
```

## Storage Class features

> Verify if there is a storage class installed in the kubernetes cluster. In this example we will use the `aws-ebs-csi-driver` as storage class reference.

Reviewer: ```suggestion
Verify if there is a storage class installed in the Kubernetes cluster. In this example we will use the `aws-ebs-csi-driver` as the storage class reference.
```

```
> kubectl get sc
NAME                             PROVISIONER       AGE
awsebscsiprovisioner (default)   ebs.csi.aws.com   2d
```

### Volume Expansion

> Verify if  the storage class has the option `AllowVolumeExpansion` to `true` 

Reviewer: ```suggestion
Verify if the storage class has the option `AllowVolumeExpansion` and is set to `true`.
```

```
> kubectl describe sc awsebscsiprovisioner
Name:                  awsebscsiprovisioner
IsDefaultClass:        Yes
Annotations:           kubernetes.io/description=AWS EBS CSI provisioner StorageClass,storageclass.kubernetes.io/is-default-class=true
Provisioner:           ebs.csi.aws.com
Parameters:            type=gp2
AllowVolumeExpansion:  true
MountOptions:          <none>
ReclaimPolicy:         Delete
VolumeBindingMode:     WaitForFirstConsumer
Events:                <none>
```

> :warning: In case it's `unset` or `false`, make sure to provision enough disk when bootstrapping the KUDO Kafka cluster. The disk size can be configured using the `DISK_SIZE` parameter. By **default, its 5Gi** and **not ideal for production usage**. Extend it as much as you consider your Kafka cluster needs to have reliable stability.​ 

Reviewer: ```suggestion
:warning: In case `AllowVolumeExpansion` is `unset` or `false`, make sure to provision enough disk when bootstrapping the KUDO Kafka cluster. The disk size can be configured using the `DISK_SIZE` parameter. By **default, DISK_SIZE is set to 5Gi** and is **not ideal for production usage**. Users should Increase disk size by as much as they deem necessary for reliable stability.
```

### ReclaimPolicy

> Verify if the storage class has the option `ReclaimPolicy` set to `Retain`

Reviewer: ```suggestion
Verify the storage class has the option `ReclaimPolicy` and is set to `Retain`.
```

> To read more about the `ReclaimPolicy` check the Kubernetes official docs around [Changing Reclaim Policy](https://kubernetes.io/docs/tasks/administer-cluster/change-pv-reclaim-policy/)

Reviewer: ```suggestion
To read more about the `ReclaimPolicy` read the official Kubernetes docs on [Changing the Reclaim Policy](https://kubernetes.io/docs/tasks/administer-cluster/change-pv-reclaim-policy/)
```

```

> PersistentVolumes can have various reclaim policies, including “Retain”, “Recycle”, and “Delete”. For dynamically provisioned PersistentVolumes, the default reclaim policy is “Delete”. This means that a dynamically provisioned volume is automatically deleted when a user deletes the corresponding PersistentVolumeClaim. This automatic behavior might be inappropriate if the volume contains precious data. In that case, it is more appropriate to use the “Retain” policy. With the “Retain” policy, if a user deletes a PersistentVolumeClaim, the corresponding PersistentVolume is not be deleted. Instead, it is moved to the Released phase, where all of its data can be manually recovered.

Reviewer: Let's put [this line](https://github.com/kudobuilder/operators/pull/101/files#diff-65f62200f453450481f7bbf3dc37ae9cR49) under this section since this body of text covers the general information and the link will direct users for more advanced information.

> PersistentVolumes can have various reclaim policies, including “Retain”, “Recycle”, and “Delete”. For dynamically provisioned PersistentVolumes, the default reclaim policy is “Delete”. This means that a dynamically provisioned volume is automatically deleted when a user deletes the corresponding PersistentVolumeClaim. This automatic behavior might be inappropriate if the volume contains precious data. In that case, it is more appropriate to use the “Retain” policy. With the “Retain” policy, if a user deletes a PersistentVolumeClaim, the corresponding PersistentVolume is not be deleted. Instead, it is moved to the Released phase, where all of its data can be manually recovered.

Reviewer: ```suggestion
> PersistentVolumes can have various reclaim policies, including “Retain”, “Recycle”, and “Delete”. For dynamically provisioned PersistentVolumes, the default reclaim policy is “Delete”. This means that a dynamically provisioned volume is automatically deleted when a user deletes the corresponding PersistentVolumeClaim. This automatic behavior might be inappropriate if the volume contains precious data. In that case, it is more appropriate to use the “Retain” policy. With the “Retain” policy, if a user deletes a PersistentVolumeClaim, the corresponding PersistentVolume is not be deleted. Instead, it is moved to the Released phase, where all of its data can be manually recovered.
```
Let's make this a quote rather than code snippet

```

> If the `StorageClass` is shared between many users, its a common practice to keep the default `ReclaimPolicy` as `Delete` and the alternative is to change the `ReclaimPolicy` in the `PersistentVolume` once the cluster is up and running. 

Reviewer: ```suggestion
If the `StorageClass` is to be shared between many users, a common practice is to leave the default `ReclaimPolicy` as `Delete` and set `ReclaimPolicy: Retain` in the `PersistentVolume` once the cluster is up and running. 
```

> Lets see an example of KUDO Kafka 3 brokers cluster's `PersistentVolumes` where the `StorageClass` default `ReclaimPolicy` is `Delete` 

Reviewer: ```suggestion
Let's see an example of a 3-broker KUDO Kafka cluster's `PersistentVolumes` where the `StorageClass` default `ReclaimPolicy` is `Delete` 
```

pvc-8602a698-14a0-4c3d-85e6-67eb6da80a5d   5Gi        RWO            Delete           Bound    default/kafka-datadir-kafka-kafka-2                                awsebscsiprovisioner            18h
pvc-de527673-8bee-4e38-9e6d-399ec07c2728   5Gi        RWO            Delete           Bound    default/kafka-datadir-kafka-kafka-1                                awsebscsiprovisioner            18h
```

We can patch the `PersistentVolumes` to use the `ReclaimPolicy` of `Retain`

```
kubectl patch pv pvc-6a9e69f4-b807-440f-b190-357c109e8ad9 -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
kubectl patch pv pvc-8602a698-14a0-4c3d-85e6-67eb6da80a5d -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
kubectl patch pv pvc-de527673-8bee-4e38-9e6d-399ec07c2728 -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'

persistentvolume/pvc-6a9e69f4-b807-440f-b190-357c109e8ad9 patched
persistentvolume/pvc-8602a698-14a0-4c3d-85e6-67eb6da80a5d patched
persistentvolume/pvc-de527673-8bee-4e38-9e6d-399ec07c2728 patched
```

> Verify that `ReclaimPolicy` has been changed for the broker `PersistentVolumes`

Reviewer: ```suggestion
Verify the `PersistentVolumes` `ReclaimPolicy` has been changed for all brokers:
```

```
> kubectl get pv
pvc-6a9e69f4-b807-440f-b190-357c109e8ad9   5Gi        RWO            Retain           Bound    default/kafka-datadir-kafka-kafka-0                                awsebscsiprovisioner            18h
pvc-8602a698-14a0-4c3d-85e6-67eb6da80a5d   5Gi        RWO            Retain           Bound    default/kafka-datadir-kafka-kafka-2                                awsebscsiprovisioner            18h
pvc-de527673-8bee-4e38-9e6d-399ec07c2728   5Gi        RWO            Retain           Bound    default/kafka-datadir-kafka-kafka-1                                awsebscsiprovisioner            18h
```

## Broker configuration

#### Replication Factor

> Change the default replication factor from the default 1 to 3

Reviewer: ```suggestion
Change the default replication factor from the default `1` to `3`.
```

```
DEFAULT_REPLICATION_FACTOR=3
```

#### Minimum Insync Replicas

> When a producer sets acks to `all` (or `-1`) `MIN_INSYNC_REPLICAS` is the minimum number of replicas that must acknowledge a write for the write to be considered successful.

Reviewer: ```suggestion
`MIN_INSYNC_REPLICAS` is the minimum number of replicas that must acknowledge a write for the write to be considered successful. To enforce greater durability guarantees, Kafka users should use `MIN_INSYNC_REPLICAS` in conjunction with producer `acks`.
```

> When used together, `min.insync.replicas` and `acks` allow Apache Kafka users to enforce greater durability guarantees.

Reviewer: ```suggestion
```

> To guarantee the messages durability a recommended practice would be to create a topic with a replication factor of `3`, set `MIN_INSYNC_REPLICAS` to `2`, and produce messages with acks of `all`.

Reviewer: ```suggestion
To guarantee message durability, a recommended practice is to create a topic with a replication factor of `3`, set `MIN_INSYNC_REPLICAS` to `2`, and produce messages with acks of `all`.
```

```
MIN_INSYNC_REPLICAS=2
```

#### Number of partitions

> The number of partitions should be higher than `1` for reliability reasons, by default this option is set to `3`

Reviewer: ```suggestion
The number of partitions should be higher than `1` for reliability reasons. By default, this option is set to `3`.
```

> More partitions lead to higher throughput performance but affect latency and availability. In KUDO Kafka we can set this trade-off of throughput vs. availability by setting the default partition number.

Reviewer: ```suggestion
More partitions lead to higher throughput performance but affect latency and availability. In KUDO Kafka we can balance this trade-off of throughput vs. availability by configuring the default partition number.
```

```
NUM_PARTITIONS=3
```

#### Graceful rolling restarts

> KUDO Kafka enables by default the option `CONTROLLED_SHUTDOWN_ENABLE` to true. To ensure a faster flush during shutdown and faster recovery during the bootstrap `num.recovery.threads.per.data.dir` can be increased from default `1` thread.

Reviewer: ```suggestion
By default, KUDO Kafka sets `CONTROLLED_SHUTDOWN_ENABLE` to `true`. To ensure a faster flush during shutdown, and faster recovery during bootstrap, `num.recovery.threads.per.data.dir` can be increased from the default of `1` thread.
```

I'm a little lost on how `CONTROLLED_SHUTDOWN_ENABLE` came into this picture when discussing recovery.threads. Perhaps provide a small explanation on controlled_shutdown first?

Author: ok let me add a bit more explanation here, if `CONTROLLED_SHUTDOWN_ENABLE` is `false` there is no faster flush as the brokers will be just killed.

```
NUM_RECOVERY_THREADS_PER_DATA_DIR=3
```

#### Network Threads

> If the Kafka cluster will have a lot of income requests, then you might need to change from the default value of `3` threads. This is necessary for clusters where a lot of different producers are connecting and writing messages.

Reviewer: ```suggestion
If users expect the Kafka cluster to receive a high number of incoming requests, then users may need to increase the the number of network threads from its default value of of `3`. This is important for clusters where a lot of different producers are connecting and writing messages.
```

#### Graceful rolling restarts

> By default, KUDO Kafka sets `CONTROLLED_SHUTDOWN_ENABLE` to `true`. Which means whenever the pods are restarted the broker does a controlled shutdown. That process includes a log flush. During these controlled shutdown and bootstrap periods Apache Kafka's LogManager uses a Threadpool. To ensure a faster flush during shutdown, and faster recovery during bootstrap, we can configure the threads in the LogManager Threadpool using the property `num.recovery.threads.per.data.dir`. Which can be increased from the default value of `1` thread.

Reviewer: 👍
